compiler/rustc_ty_utils/src/consts.rs

```diff
@@ -117,7 +117,11 @@ fn recurse_build<'tcx>(
         }
         &ExprKind::Literal { lit, neg } => {
             let sp = node.span;
-            match tcx.at(sp).lit_to_const(LitToConstInput { lit: &lit.node, ty: node.ty, neg }) {
+            match tcx.at(sp).lit_to_const_for_patterns(LitToConstInput {
```

Comment:
this is definitely not a pattern xd we'd end up here for things like the `1` in `foo::<{ N + 1 }>();`